### ISSUE TITLE
Localized routes

### ISSUE THREAD

**Issue by exophunk** [s00d/nuxt-i18n-micro#6]
I'm curious about switching to nuxt-i18n-micro as nuxt-i18n is a bit clunky. I am almost sold, but I am hesitating because of some features I'm unsure:

## Localized routes
Is it possible, or will it be possible at some point to have localized route paths? Like `products/apple` and `produkt/apfel`

## Disable page-based translations
I like the straight-forward approach, but in our case, we have many similiar pages that use shared components. As you already support layers, it is sufficient for us to have translations separated per layer. Can i disable page-based translations?

## Localize dates/numbers etc.
When I understand correctly, you're not using vue-i18n right? I was used to fallback to $n() etc. for localized number formats. What would a good approach be with nuxt-18n-micro? Should I use `Intl.NumberFormat` with separate own helper utilities or so?

**Comment by s00d** [comment]
Hello 

> Localized routes

I haven't implemented this functionality yet, and in our case, it doesn't make much sense. We use slugs for this and fetch content from the backend. We need to see how this can be done more easily. However, since this is quite a complex functionality, I will need some time for implementation and testing.

> Disable page-based translations

There's no such feature available either, but it's easy to add. I'll try to add it soon. In principle, you can already use just one file and ignore the others, but this would lead to longer loading times and more complex page rendering.

>Localize dates/numbers, etc.

I intentionally didn't implement such functionality because dates generally don't fall under i18n and localization. For this purpose, there are excellent libraries such as [Luxon](https://moment.github.io/luxon/#/intl). Luxon is, in any case, far superior to anything you can do with i18n.

**Comment by s00d** [comment]
Added the ability to disable per-page localization: [docs](https://s00d.github.io/nuxt-i18n-micro/guide/getting-started#%F0%9F%94%84-disablepagelocales-boolean)

Added localized routes: [docs](https://s00d.github.io/nuxt-i18n-micro/guide/custom-locale-routes)

**Comment by exophunk** [comment]
wow awesome and fast work! I hope you even got some sleep :-)
I am testing it right away and will give you feedback. looks very promising. I found a little issue, probably unrelated. I'll open a separate issue for that.

**Comment by exophunk** [comment]
I have tried to setup nuxt-micro-i18n and run into some issues. It seems I may have some basic config issue, as translations do not work and I have some trouble with routing. I am not sure If this is related to the latest 1.11 release, but I also seem to can't get it to work on older versions. Maybe I have one tiny config issue that cause all the other problems.

I have a reproduction here to check out: https://stackblitz.com/edit/nuxt-starter-xfjjgc?file=app.vue

# Issue 1: No translations
`{{ $t('generic.welcome') }}` does not output translated strings. It seems to somehow have to to with "defaultLocale" as when I disable it, it sometimes work and show translations, but only after reload. It also does not translate after vue router navigation or locale switch. Maybe something with my basic setup is faulty, but I can't figure it out? It seems that the translation files loaded (network tab) are empty. I would need it to work also without SSR.

# Issue 2: Nested routes cause trouble
I use nested routes in our app like this:
```md
- activity.vue
- activity/
  - skiing.vue
```

Trying to create links to those subpages like `<i18n-link :to="{ name: 'activity-skiing'}">Link</i18n-link>` cause Err 500. (Same for `$localeRoute()`). This seems not to be the case on other, non-nested routes.


# Issue 3: `<i18n-switcher>` Bug on routes with custom path
This is probably all related to the same routing issues I have, but when using the switcher on a page with custom routes (see example "subpage"), the switcher does not work and break when trying to open the dropdown.

# Issue 4: `$defineI18nRoute()` validation
When using this method to define `localeRoutes`, it [throws an error ](https://github.com/s00d/nuxt-i18n-micro/blob/425977df6cbf0b4b4d479240cdb898ba07e2458b/src/module.ts#L413)when `locales` is not defined.

```js
$defineI18nRoute({
  localeRoutes: {
    ru: '/localesubpage', // Custom route path for the Russian locale
    de: '/lokaleseite',   // Custom route path for the German locale
  },
})
```
--> not possible!



# Suggestions
This is just my personal optionion, but I have a suggestion for two adjustments
- Maybe you can add a config to disable the file watcher on dev alltogether, so it does not create missing locale files
- I would prefer an `<i18n-switcher>` that is completely unstyled by default and allows adding styles (or classes) when desired. Even with passing custom styles, it is not possible to remove all styles.

I've built my own switcher (which is super easy, fair enough)
```vue
<button
    v-for="locale in $getLocales()"
    :key="locale.code"
    :disabled="locale.code === $getLocale()"
    @click="$switchLocale(locale.code)"
>
    {{ locale.code }}
</button>
```


Once again, thank you very much for your effort and your very fast reply to my requests yesterday :) I appreciate it.

**Comment by s00d** [comment]
I just completely forgot how the router works in Nuxt and that it can create child routes. I think I’ve finished everything, and also fixed a few other bugs.

I've also added what you [requested](https://s00d.github.io/nuxt-i18n-micro/guide/getting-started#%F0%9F%91%80-disablewatcher-boolean).

Everything you wrote should work now, but I might have missed something. It would be best to double-check it again.

Regarding the i18n-switcher, it's primarily intended as an example for testing purposes, and if we remove the styles completely, it almost loses its purpose. The best approach would be to either create a custom wrapper in the project for styling or simply write your own. The i18n-switcher can then be used as an example.

**Comment by exophunk** [comment]
Awesome work! 

As far as I checked in my example repo, everything works flawlessly now. I will now start adopting it in our main project and see if I find something else.

Thank you for your effort.

**Comment by exophunk** [comment]
Sorry to bother you again, but I still have an issue with custom routes for nested routes.

I've created an updated example: https://stackblitz.com/edit/nuxt-starter-vqutwr?file=app.vue 

Custom routes work well for normal pages. In case of parent/child routes. I would like to fully translate the whole route. For example:

```
- activities
  - skiing-page
  - hiking-page
```

should become:

EN: `book-activity/skiing`
DE: `aktivitaet-buchen/ski-fahren`


Parent:
```js
$defineI18nRoute({
  localeRoutes: {
    en: '/book-activity',
    de: '/aktivitaet-buchen',
  },
});
```

Child
```js
$defineI18nRoute({
  localeRoutes: {
    en: '/skiing',
    de: '/ski-fahren',
  },
});
```

I've tried various combinations of specifying $defineI18nRoute() in both parent/child. I also tried to include parent segmment in child too etc. I've end up that either the parent or the child path is not translated (they exist both), e.g.

`activity/aktivitaet-buchen/skifahren` (only child part translated)
`aktivitaet-buchen/skiing-page` (only parent part translated)

but no "fully" translated route is created. I've studied the code and see you're incoporating parent/child relations. Maybe my config are wrong? see the example.



Another tiny issue I found with $defineI18nRoute() is that the [regex extraction](https://github.com/s00d/nuxt-i18n-micro/blob/main/src/module.ts#L454) also happens when commented.

So 
```js
// $defineI18nRoute({
//     localeRoutes: {
//         en: '/ignore...',
//         de: '/ignore...',
//     },
// })
```

is still matched and the extractor code then breaks

**Comment by s00d** [comment]
### Try the New Version

[Nuxt-i18n-micro v1.12.2](https://github.com/s00d/nuxt-i18n-micro/releases/tag/v1.12.2)

Working with custom routes turned out to be quite complex. I had to write numerous tests to cover all possible link formation scenarios. While I believe I've tested everything, there might still be edge cases where things could break. It's essential to check how it performs now.

Honestly, I was reluctant to implement these custom routes for a reason—there's just too much unnecessary logic involved. Working with slugs is much simpler and more intuitive. 😅

Additionally, I’ve reworked the defineI18nRoute search. Now, it will skip commented lines.

**Comment by s00d** [comment]
I rewrote the entire logic for generating links; it's now encapsulated in a separate class, and I added tests for it. Everything seems to be working correctly.

**Comment by exophunk** [comment]
I have the following issue:

```vue
{{ $localeRoute({ name: 'checkout-step-2-booking', params: { bookingId: '123' } }) }}
```

throws an error: `Error: Missing required param "bookingId"`

Folders:
```
-checkout/
  - step-2-booking.vue
```

so normal route name is: `checkout-step-2-booking`.

without i18n I define 
```
definePageMeta({
    path: '/checkout/booking/:bookingId',
 });
 ```
 
 but even if I use nuxt defaults and name the page: `step-2-booking-[bookingId]`
 
 it seems to be an issue with the $localeRoute() composable not handling params correctly? Or maybe my page/params setup is off.
 
 
Everything else works fine! 👍
 
  Some more suggestions:

- I've created a manual helper function `useBaseRouteName()` that returns route name without i18n stuff. I see you also have a [helper](https://github.com/s00d/nuxt-i18n-micro/blob/main/src/runtime/plugins/01.plugin.ts#L35) for this and use it frequently. You should expose it as well? I use a lot of `navigateTo($localeRoute({ name: 'route-name' }))` and I constantly have to normalize route names everywhere. For example `useRoute().name` of the current route returns the localized version. Working with Routes, I assume almost everyone usually just wants to navigate to a specific page and not a specific locale. So using your helper instead of creating a custom one.

- The return type of `$t()` is not string, but `string | number | boolean | Translations | PluralTranslations | unknown[] | unknown | null`. Maybe I dont understand the implications, but doesnt $t() always return a string, even if it doesnt find the key? Because now I have to explitly say `const label = $t('key') as string` everywhere to prevent type errors (when i expect string). maybe i'm doing it wrong, I dont have that much TS experience. But having to add `as string` everywhere seems annoying.

**Comment by s00d** [comment]
> I've created a manual helper function useBaseRouteName() that returns route name without i18n stuff. I see you also have a [helper](https://github.com/s00d/nuxt-i18n-micro/blob/main/src/runtime/plugins/01.plugin.ts#L35) for this and use it frequently. You should expose it as well? I use a lot of navigateTo($localeRoute({ name: 'route-name' })) and I constantly have to normalize route names everywhere. For example useRoute().name of the current route returns the localized version. Working with Routes, I assume almost everyone usually just wants to navigate to a specific page and not a specific locale. So using your helper instead of creating a custom one.


I added a public method: [getRouteName](https://s00d.github.io/nuxt-i18n-micro/api/methods#%F0%9F%94%8D-getroutename-route-routelocationnormalizedloaded-routelocationresolvedgeneric-locale-string-string).

> The return type of $t() is not string, but string | number | boolean | Translations | PluralTranslations | unknown[] | unknown | null. Maybe I dont understand the implications, but doesnt $t() always return a string, even if it doesnt find the key? Because now I have to explitly say const label = $t('key') as string everywhere to prevent type errors (when i expect string). maybe i'm doing it wrong, I dont have that much TS experience. But having to add as string everywhere seems annoying.


No, the issue you're experiencing is actually one of the main problems with traditional i18n libraries, which is why I chose not to use them. In my implementation, you can comfortably receive an object, array, integer, or almost any data type, which greatly simplifies working with different types of data, such as displaying a list. If you provide an example, I can try to adjust the typing to address this problem.

For the first part, it's hard for me to say anything specific. I need an example to understand your logic. It's better to create a separate issue.

**Comment by exophunk** [comment]
Thank you for your explanation. I see your point with the types. And thanks for adding getRouteName
I create a second issue for the params.
